File: 002/solution.rb
#!/usr/bin/env ruby
fibonacci = Enumerator.new do |result|
  a = b = 1
  loop do
    result << a
    a, b = b, a + b
  end
end

maximum = 4_000_000
puts fibonacci.take_while { |x| x < maximum }.select { |x| x % 2 == 0 }.reduce :+

File: 002/solution.py
#!/usr/bin/env python
from itertools import ifilter, takewhile

def fibonacci(start=1):
    """Generates an infinite length Fibonacci sequence."""
    a, b = 0, start
    while True:
        c = a + b
        yield c
        a, b = b, c

def is_even(n):
    """Returns True if a number is even."""
    return n % 2 == 0

maximum = 4 * 1e6
print sum(takewhile(lambda x: x < maximum, ifilter(is_even, fibonacci())))
